scripts/collect_activations.py

```diff
@@ -1,77 +1,169 @@
+        layer: [] for layer in layers_to_cache
+    }
+
+    tokenized_instruction = tokenizer.encode(
+        f"\n\n{INSTRUCTION_STR}\n", add_special_tokens=False, return_tensors="pt"
+    )[0]
+
+    for q_idx, question in tqdm(enumerate(eval_questions)):
+        tokenized_fsp = torch.tensor(fsps[q_idx])
+        tokenized_question = torch.tensor(question.tokenized_question)
+        tokenized_response = torch.tensor(tokenized_responses[q_idx])
+
+        # Build input tokens
+        input_ids = torch.cat(
+            (
+                tokenized_fsp,
+                tokenized_question,
+                tokenized_instruction,
+                tokenized_response,
+            ),
+            dim=0,
+        )
 
-    model = AutoModelForCausalLM.from_pretrained(model_name).cuda()
+        locs_to_cache = []
+        for loc_type in loc_types_to_cache:
+            if loc_type == "resp":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp)
+                        + len(tokenized_question)
+                        + len(tokenized_instruction),
+                        len(input_ids),
+                    )
+                )
+            elif loc_type == "instr":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp) + len(tokenized_question),
+                        len(tokenized_fsp)
+                        + len(tokenized_question)
+                        + len(tokenized_instruction),
+                    )
+                )
+            elif loc_type == "quest":
+                locs_to_cache.extend(
+                    range(
+                        len(tokenized_fsp), len(tokenized_fsp) + len(tokenized_question)
+                    )
+                )
+            elif loc_type == "fsp":
+                locs_to_cache.extend(range(len(tokenized_fsp)))
 
-    layers_to_cache = list(range(model.config.num_hidden_layers))
-    activations_by_question: list[QuestionActivations] = []
-    for question in tqdm(questions):
-        locs_to_cache = set()
-        for key, locs in question.locs.items():
-            locs_to_cache.update(locs)
-        locs_to_cache = sorted(locs_to_cache)
-
-        # # hack to add "Let's think step by step"
-        # min_loc = locs_to_cache[0]
-        # # hack to not add anything else
-        # locs_to_cache = list(range(min_loc - 10, min_loc))  # + locs_to_cache
-        # # hack to have smaller prompt and faster inference
-        # input_ids = torch.tensor(question.tokens[:min_loc])
-        input_ids = torch.tensor(question.tokens)
         resid_acts = clean_run_with_cache(
             model, input_ids, layers_to_cache, locs_to_cache
         )
 
-        activations_by_question.append(QuestionActivations(resid_acts, locs_to_cache))
+        for layer_idx, layer in enumerate(layers_to_cache):
+            activations_by_layer[layer].append(resid_acts[layer_idx].cpu())
```

Comment:
clean_run_with_cache should just return list

Reply:
I already need to iterate over layers in `process_questions`, so I will do it just there